compiler/passes/denormalize.cpp

```diff
@@ -669,3 +676,259 @@ inline bool unsafeExprInBetween(Expr* e1, Expr* e2, Expr* exprToMove,
+    return false;
+
+  INT_ASSERT(call == call->getStmtExpr());
+  SymExpr* tempSE = toSymExpr(defInfo.skipLnFnArgs ?
+                      call->argList.tail->prev->prev : call->argList.tail);
+  // the temp is usually called "ret_tmp", however not necessarily
+  INT_ASSERT(tempSE->symbol()->type == defInfo.refArg->type); // fyi
+
+  useInfo = { fnUse, tempSE };
+  return true;
+}
+
+static void transformRetTempUse(UseInfoRR& info) {
+  Expr*   fnCall  = info.fnSE->parentExpr;
+  Symbol* retTemp = info.tempSE->symbol();
+  //fyi:
+  INT_ASSERT(! retTemp->type->symbol->hasEitherFlag(FLAG_REF, FLAG_WIDE_REF));
+
+  // replace:
+  //   call fn(args, ret_tmp)
+  // with:
+  //   move(ret_tmp, call fn(args))  // if ret_tmp is a ref, make it assign
+  //
+  SET_LINENO(fnCall);
+
+  Expr* anchor = fnCall->prev;
+  BlockStmt* encl = anchor ? NULL : toBlockStmt(fnCall->parentExpr);
+  Expr* move   = new CallExpr(retTemp->isRef() ? PRIM_ASSIGN : PRIM_MOVE,
+                              info.tempSE->remove(), fnCall->remove());
+  if (anchor) anchor->insertAfter(move);
+  else        encl->insertAtHead(move);
+}
+
+///////////
+// putting it all together
+
+static void undoReturnByRef(FnSymbol* fn) {
+  if (fn->hasFlag(FLAG_VIRTUAL)) return;  // skip these for now
+
+  DefInfoRR defInfo;
+  if (! acceptableDef(fn, defInfo))
+    return;
+
+  // Make changes only if we can handle all uses of 'fn'.
+  // While checking, store some findings for later use.
+  std::vector<UseInfoRR> infos;
+  for_SymbolSymExprs(use, fn) {
+    UseInfoRR useInfo;
+    if (acceptableUse(defInfo, use, useInfo))
+      infos.push_back(useInfo);
+    else
+      return;
+  }
+
+  for (UseInfoRR& info: infos)
+    transformRetTempUse(info);
+
+  transformRetTempDef(defInfo);
+}
+
+/////////////////////////////////////////////////////////////////////////////
+// collapseTrivialMoves() converts:
+//   move(source, expr)  // move1
+//   move(dest, source)  // move2
+// provided:
+//   'move1' and 'move2' are adjacent or have only DefExprs in between
+//   'source' has no other references
+// to:
+//   move(dest, expr)
+
+// In some cases codegen adds an appropriate dereference, widening, etc.
+// for a symbol-to-symbol move and not for a call-to-symbol move.
+// So collapse only moves for which such additions are not needed.
+static bool okSymbol(Symbol* sym) {
+  return sym->qual == QUAL_VAL || sym->qual == QUAL_CONST_VAL;
+}
+
+static bool okToCollapse(SymExpr* dest, SymExpr* source) {
+  Symbol *destSym = dest->symbol(), *sourceSym = source->symbol();
+  return
+    destSym->type == sourceSym->type        &&
+    ! sourceSym->hasFlag(FLAG_CONFIG)       &&
+    ! sourceSym->hasFlag(FLAG_EXPORT)       &&
+    ! sourceSym->hasFlag(FLAG_EXTERN)       &&
+    okSymbol(destSym) && okSymbol(sourceSym);
+}
+
+static bool closeEnough(Expr* move1, Expr* move2) {
+  Expr* curr = move1;
+  for (int dist = 0; dist < 5; dist++) {  // heuristically allow <=5 defs
+    curr = curr->next;
```

Comment:
What happens if you don't have `closeEnough`? Does stuff break, or are you just being careful?

Reply:
I have not convinced myself that there will never be intervening moves or function calls that will make collapsing invalid.